Adding FastAI(pytorch) as a first class citizen to the pipeline

adricl opened the pull request:
This PR allows you to use Fast AI(pytorch) end to end in donkey car. I have tested training and driving with the simulator.

I have added a new type of model fastai_ so far linear is the only model implemented. More can be added in the future of course.
I have tried my best to emulate most of the function that Keras does but there are some idiosyncrasies between the two frameworks. 

These normal training and driving commands work.
python train.py --tubs data/ --model models/fastai_linear_model.pth --type fastai_linear
python manage.py drive --model models/fastai_linear_model.pth --type fastai_linear

I do realise that this PR comes without a heads up or warning so if its rejected that's fine. 
I do really like this project and the work that has been done, its just I prefer PyTorch to TensorFlow. 

Any suggestions are welcome. I am new to python of this complexity so any pointers would be great.
There will be more cleanups and fixes on this PR as I am not finished yet. I need to add some graph plotting and other tidy ups and maybe a refactor. Any guidance would be great. 
I will also add some docs around this.

If this goes well I might also look at getting pytorch lightning into own part and working in the pipeline.

Please note this will not work with TensorRT or TF lite if there is demand I could look at this in the future.

Cheers

DocGarbanzo commented:
@adricl - are you still working on this PR? It will take a bit to go through this. You should definitely make sure that the tests are passing. If you are still working on it, then you can convert the PR to `draft`.

adricl commented:
> Contributor

Thanks for that, I know its a big piece of work. Sorry about that. 
I am trying to get the tests to pass. I have converted it to draft,.

tikurahul commented:
One more thing. When you are ready, can you please squash your changes to a single commit ? That way it becomes easier to reason about.

adricl commented:
@tikurahul to squash the commit do I have to create a new PR or can I do it in this existing PR? I am not sure how to go about it if you can do it for the existing pr. Thanks

adricl commented:
I think this pr is good to review now..

DocGarbanzo commented:
@adricl - you don't need to do a new PR. You can squash the commits on your branch, make sure everything is working and force-push. It will update this PR here in-place.

DocGarbanzo commented on donkeycar/parts/interpreter.py:
```diff
@@ -165,6 +169,70 @@ def load_weights(self, model_path: str, by_name: bool = True) -> \
+        self.model.eval()
+
+    def summary(self) -> str:
+        return self.model
```
Does this return a string for description?

adricl replied:
Yeah this does return a string, I have modified the Keras Class to return the same.

DocGarbanzo replied:
Hm... looking at lines 220 and 223, it doesn't look like `self.model` is string but rather a PyTorch model.

DocGarbanzo commented on donkeycar/parts/interpreter.py:
```diff
@@ -165,6 +169,70 @@ def load_weights(self, model_path: str, by_name: bool = True) -> \
+    def load(self, model_path: str) -> None:
+        logger.info(f'Loading model {model_path}')
+        if torch.cuda.is_available():
+            print("using cuda for torch inference")
```
Pls use logging instead of print

adricl replied:
Updated to use loggging.

DocGarbanzo commented on donkeycar/parts/interpreter.py:
```diff
@@ -165,6 +169,70 @@ def load_weights(self, model_path: str, by_name: bool = True) -> \
+            inputs = [img_arr, other_arr]
+        return self.invoke(inputs)
+
+    def predict_from_dict(self, input_dict):
```
Do you use this function?

adricl replied:
This has been removed

DocGarbanzo commented on donkeycar/parts/pytorch/torch_data.py:
```diff
@@ -41,7 +41,7 @@ def get_default_transform(for_video=False, for_inference=False):
 
 
-class TorchTubDataset(IterableDataset):
+class FastaAiTorchTubDataset(IterableDataset):
```
Looks like this class has now been defined twice, pls check comment above.

adricl replied:
This has been reverted.

DocGarbanzo commented on donkeycar/pipeline/training.py:
```diff
@@ -81,6 +81,14 @@ def create_tf_data(self) -> tf.data.Dataset:
             output_shapes=self.model.output_shapes())
         return dataset.repeat().batch(self.batch_size)
 
+    def create_tf_data(self) -> tf.data.Dataset:
```
Please remove, this method is already defined 8 lines above.

adricl replied:
So sorry about this.

DocGarbanzo commented on donkeycar/pipeline/training.py:
```diff
@@ -162,12 +179,12 @@ def train(cfg: Config, tub_paths: str, model: str = None,
     database.write()
 
-    return history
+    return history
```
EOF.

adricl replied:
Done

DocGarbanzo replied:
It still shows missing newline at EOF.

DocGarbanzo commented on donkeycar/utils.py:
```diff
@@ -428,15 +428,17 @@ def eprint(*args, **kwargs):
 
 
-def get_model_by_type(model_type: str, cfg: 'Config') -> 'KerasPilot':
+def get_model_by_type(model_type: str, cfg: 'Config') -> 'ABC':
```
You can return `Union['KerasPilot', 'FastAiPilot']`

adricl replied:
Done

DocGarbanzo commented on donkeycar/parts/interpreter.py:
```diff
@@ -3,15 +3,16 @@
 import logging
 import numpy as np
 from typing import Union, Sequence, List
+from pathlib import Path
```
Are you using this?

adricl replied:
No this was a mistake

DocGarbanzo commented on donkeycar/utils.py:
```diff
@@ -428,15 +428,17 @@ def eprint(*args, **kwargs):
         KerasInferred, KerasIMU, KerasMemory, KerasBehavioral, KerasLocalizer, \
         KerasLSTM, Keras3D_CNN
-    from donkeycar.parts.interpreter import KerasInterpreter, TfLite, TensorRT
+    from donkeycar.parts.interpreter import KerasInterpreter, TfLite, TensorRT, FastAIInterpreter
```
Pls keep pep-8 compliant 80 char lines.

adricl replied:
Fixed.

DocGarbanzo reviewed (changes requested):
It's a great idea to bring in fastai. A couple of points need to be addressed. Please also add a test for the fastai model - you can compare how @EricWiener adapted the Keras tests to PyTorch. We should test the functionality of the model and the pipeline and the training. The latter should be switched off in the CI (we have the same setup also for TF and PyTorch).

adricl commented:
Sorry I was squashing the commits and I Github closed the PR I have created a new one in #986 986 I will address your issues there. So sorr

adricl commented on donkeycar/parts/fastai.py:
```diff
@@ -1,31 +1,406 @@
+
+        self.learner = Learner(dataLoader, model, loss_func=self.loss, path=Path(model_path).parent)
+
+        print(self.learner.summary())
```
Updated

adricl commented on donkeycar/parts/fastai.py:
```diff
@@ -1,31 +1,406 @@
+        history = list(map((lambda x: x.item()), self.learner.recorder.losses))
+
+        return history
+
```
I have removed these functions.

adricl commented on donkeycar/parts/fastai.py:
```diff
@@ -1,31 +1,406 @@
+        throttle = interpreter_out[1]
+        print(steering, throttle)
         return steering, throttle
-    
```
I am not sure if these have been removed now. I am on linux so this is quite odd.

DocGarbanzo replied:
Maybe I'm misinterpreting the git diff here. Should be all good I guess.

adricl commented on donkeycar/parts/fastai.py:
```diff
@@ -1,31 +1,406 @@
+
+logger = getLogger(__name__)
+#TODO change this to use fastai Transforms
+def get_default_transform():
```
I have fixed this so we don't have duplicate code. I did this as I thought the changes I would be doing were going to be huge, they turned out to be so minor.

In the future(not this PR) I will make use of some of the other transforms that fastai provides or albumentations if you guys implement it.

adricl commented:
I have to look at the tests next. That might take me a few days.

DocGarbanzo commented on install/envs/ubuntu.yml:
```diff
@@ -36,6 +37,7 @@ dependencies:
   - plotly
   - pyyaml
   - tensorflow=2.2.0
+  - fastai
```
Please also add the module in `setup.py`. Do you know if this installs on the RPi w/o problems?

adricl replied:
I think it should work as fastai is a wrapper around pytorch so there is no binaries as such. I will test this.

DocGarbanzo commented:
@adricl - this looks much better now.

Bleyddyn commented:
Thanks for this. I had just started working on the same idea but I haven't gotten nearly as far.
I have a notebook that has a working DataBlock for loading v2 Tub files, but I'm still trying to figure out how to train using it: https://github.com/Bleyddyn/malpi/tree/master/notebooks

adricl commented:
@Bleyddyn off topic but this is how it was done in Tub V1 [link](https://github.com/cmasenas/fastai_navigation_training/blob/master/fastai_train.ipynb) This was a very clean implementation. I do have created a notebook so we can experiment with models but its not releasable as yet. I will have to rewrite it for this pr.

DocGarbanzo commented:
@adricl @Bleyddyn - if you want to put records into a data frame as in the example above, this is now much easier with the new tub, because it is an Iterable over dictionaries. You just say:
```python
import pandas as pd
from donkeycar.parts.tub_v2 import Tub
t = Tub('/home/me/mycar/data')
df = pd.DataFrame(t)
```

Bleyddyn commented:
@DocGarbanzo That's what I'm doing in the notebook I linked to. Updated now with a better DataBlock definition and actual working training code.

DocGarbanzo commented:
@Bleyddyn - just looked at your notebook now. Very good.

adricl commented:
@DocGarbanzo I have updated the test train pipeline to run fastai. I have also added fastai to the pi build. Is there anything else that needs to be completed for this PR?

DocGarbanzo commented on donkeycar/tests/test_train.py:
```diff
@@ -13,7 +13,7 @@
 Data = namedtuple('Data',
-                  ['type', 'name', 'convergence', 'pretrained', 'preprocess'],
+                  ['type', 'name', 'convergence', 'pretrained', 'preprocess', 'tf_lite', 'tensor_rt'],
                   defaults=(None, ) * 5)
```
If you add 2 more arguments to the namedtuple, please also increase the default initialisation by 2, i.e. 5 -> 7.

adricl replied:
This is now resolved

DocGarbanzo reviewed (changes requested):
This is very good now, can you please address the small comment on the test & push it, then I'll merge the code.

adricl commented:
We are good to go!

c1505 commented:
Awesome contribution here !  I would like to try this out in the simulator and/or on a car.  Any docs associated with this update or plan to create/modify some ? I can help review/validate the doc changes . @adricl

adricl commented:
Thanks,
I will have to add documentation at some stage. 
Basically you need to train and load with model_type = "fastai_linear". This will will work end to end with the simulator as that is where I was running it. 
I will get onto the docs asap.

adricl commented:
@c1505  this is my documentation. [FastAi Docs](https://github.com/adricl/donkeydocs/blob/master/docs/parts/fastai.md) Have a look before I merge it.
